File: src/2025-03-04-03-28-40--deepseek-chat-0304/java/exercises/practice/book-store/src/main/java/BookStore.java
import java.util.List;
import java.util.Collections;
import java.util.ArrayList;

class BookStore {
    private static final double BOOK_PRICE = 8.0;
    private static final double[] DISCOUNTS = {0, 0, 0.05, 0.10, 0.20, 0.25};

    double calculateBasketCost(List<Integer> books) {
        if (books == null || books.isEmpty()) {
            return 0.0;
        }

        // Validate book numbers are between 1 and 5
        for (int book : books) {
            if (book < 1 || book > 5) {
                throw new IllegalArgumentException("Book numbers must be between 1 and 5");
            }
        }

        // Count how many of each book we have
        int[] counts = new int[5];
        for (int book : books) {
            counts[book - 1]++;
        }

        // Try different grouping strategies to find the best discount
        double total1 = calculateTotal(counts.clone(), 5);
        double total2 = calculateTotal(counts.clone(), 4);
        
        return Math.min(total1, total2);
    }

    private double calculateTotal(int[] counts, int maxGroupSize) {
        double total = 0.0;
        int[] tempCounts = counts.clone();
        
        while (true) {
            int groupSize = 0;
            int uniqueBooks = 0;
            
            // Try to form the largest possible group
            for (int i = 0; i < tempCounts.length; i++) {
                if (tempCounts[i] > 0) {
                    groupSize++;
                    uniqueBooks++;
                    tempCounts[i]--;
                }
                
                if (groupSize == maxGroupSize) {
                    break;
                }
            }
            
            if (groupSize == 0) {
                break;
            }
            
            // Calculate price for this group
            total += groupSize * BOOK_PRICE * (1 - DISCOUNTS[uniqueBooks]);
            
            // Try to form smaller groups if we couldn't make a full group
            if (groupSize < maxGroupSize) {
                maxGroupSize = groupSize;
            }
        }
        
        return total;
    }
}

File: src/2025-03-04-03-28-40--deepseek-chat-0304/python/exercises/practice/book-store/book_store.py
def _calculate_group_price(group_size):
    """Calculate price in cents for a group of unique books with appropriate discount."""
    discounts = {1: 0, 2: 0.05, 3: 0.10, 4: 0.20, 5: 0.25}
    return int(group_size * 800 * (1 - discounts.get(group_size, 0)))

def total(basket):
    """Calculate the total price of a basket of books with maximum discount."""
    if not basket:
        return 0
    
    # Count how many of each book we have
    counts = {}
    for book in basket:
        counts[book] = counts.get(book, 0) + 1
    
    total_price = 0
    while sum(counts.values()) > 0:
        # Find the largest possible group of unique books
        group_size = 0
        for book in counts:
            if counts[book] > 0:
                group_size += 1
                counts[book] -= 1
        
        # Remove books with zero count
        counts = {k: v for k, v in counts.items() if v > 0}
        
        # Calculate price for this group and add to total
        total_price += _calculate_group_price(group_size)
    
    # Check if we can get a better price by splitting a group of 5 into 3+3
    # This is needed for cases like 2,2,2,1,1 where 5+3 is worse than 4+4
    counts = {}
    for book in basket:
        counts[book] = counts.get(book, 0) + 1
    
    alt_price = 0
    while sum(counts.values()) > 0:
        # Try to make groups of 4 first
        group_size = 0
        for book in counts:
            if counts[book] > 0:
                group_size += 1
                counts[book] -= 1
            if group_size == 4:
                break
        
        # If we couldn't make a group of 4, make the largest possible group
        if group_size < 4:
            for book in counts:
                if counts[book] > 0:
                    group_size += 1
                    counts[book] -= 1
        
        # Remove books with zero count
        counts = {k: v for k, v in counts.items() if v > 0}
        
        # Calculate price for this group and add to total
        alt_price += _calculate_group_price(group_size)
    
    # Return the minimum price (maximum discount)
    return min(total_price, alt_price)
